kaldb/src/test/java/com/slack/kaldb/metadata/core/KaldbPartitioningMetadataStoreTest.java

```diff
@@ -0,0 +1,538 @@
+        ExampleMetadata toRemove = addedMetadata.remove();
+        partitionedMetadataStore.deleteAsync(toRemove);
+      }
+
+      await().until(() -> partitionedMetadataStore.listSync().size() == 50);
+      // this can be higher than 150, due to container znodes contributing events during clean
+      assertThat(counter.get()).isGreaterThanOrEqualTo(150);
+
+      for (int i = 0; i < 50; i++) {
+        ExampleMetadata toRemove = addedMetadata.remove();
+        partitionedMetadataStore.deleteAsync(toRemove);
+      }
+
+      await().until(() -> partitionedMetadataStore.listSync().size() == 0);
+      // this can be higher than 200, due to container znodes contributing events during clean
+      await().until(counter::get, (value) -> value >= 200);
+
+      await()
+          .until(
+              () -> {
+                if (curatorFramework
+                        .checkExists()
+                        .forPath("/partitioned_snapshot_listeners")
+                        .toCompletableFuture()
+                        .get()
+                    == null) {
+                  LOG.info("Parent node no longer exists");
+                  return true;
+                }
+
+                int childrenSize =
+                    curatorFramework
+                        .getChildren()
+                        .forPath("/partitioned_snapshot_listeners")
+                        .toCompletableFuture()
+                        .get()
+                        .size();
+                LOG.info("Children size - {}", childrenSize);
+                return childrenSize == 0;
+              });
+    }
+  }
+
+  @Test
+  void testAddRemoveListener() throws Exception {
+    class TestMetadataStore extends KaldbPartitioningMetadataStore<ExampleMetadata> {
+      public TestMetadataStore() {
+        super(
+            curatorFramework,
+            CreateMode.PERSISTENT,
+            new ExampleMetadataSerializer().toModelSerializer(),
+            "/partitioned_add_remove_listeners");
+      }
+    }
+
+    try (KaldbPartitioningMetadataStore<ExampleMetadata> store = new TestMetadataStore()) {
+      AtomicInteger counter = new AtomicInteger(0);
+      KaldbMetadataStoreChangeListener<ExampleMetadata> listener =
+          (testMetadata) -> counter.incrementAndGet();
+      store.addListener(listener);
+
+      await().until(() -> counter.get() == 0);
+
+      // create metadata
+      ExampleMetadata metadata1 = new ExampleMetadata("foo", "val1");
+      store.createSync(metadata1);
+
+      await().until(() -> counter.get() == 1);
+
+      assertThat(store.getSync(metadata1.getPartition(), metadata1.name)).isEqualTo(metadata1);
+      assertThat(counter.get()).isEqualTo(1);
+
+      metadata1.setExtraField("val2");
+      store.updateSync(metadata1);
+
+      await().until(() -> counter.get() == 2);
+
+      store.removeListener(listener);
+      metadata1.setExtraField("val3");
+      store.updateSync(metadata1);
+
+      Thread.sleep(2000);
```

Comment:
you wanted to add this so that the next line assert can be added with more confidence? i.e in a scenario where we actually don't remove listener but the lag in the listener triggering might cause the test to pass but it shouldn't?

Reply:
Fixed!

Reply:
Updated this with a comment for the future, but you are correct. I didn't want to have an assert immediately after, as it could just be passing due to the test running fast.